shared/utils/accessor-wrapper-types.hpp

```diff
@@ -37,54 +63,111 @@ namespace bs_hook {
         }
+
+        inline auto operator ->() const {
+            return this->operator T();
+        } 
+
+        inline auto operator *() const {
+            return this->operator T();
+        }
+
+        inline auto v() const {
+            return this->operator T();
+        }
+
+        template<typename... Targs>
+        decltype(auto) operator ()(Targs&&... args) { return this->operator T()(std::forward<Targs...>(args...)); }
+
         private:
         void* instance;
     };
 
     template<internal::NTTPString name, class T>
     struct InstanceProperty<name, T, false, true> {
-        explicit InstanceProperty(void* inst) noexcept : instance(inst) {}
+        explicit constexpr InstanceProperty(void* inst) noexcept : instance(inst) {}
         InstanceProperty& operator=(T&& t) {
             auto val = reinterpret_cast<Il2CppObject*>(instance);
             auto res = il2cpp_utils::SetPropertyValue<false>(val, name.data.data(), std::forward<decltype(t)>(t));
             if (!res) throw PropertyException(std::string("Failed to set instance property: ") + name.data.data());
             return *this;
         }
+        
         private:
         void* instance;
     };
 
+#define BINARY_OPERATOR_OP_EQ_PROP(op)      \
+template<typename U>                        \
+auto& operator op##=(const U& rhs) {        \
+    auto temp = this->operator T();         \
+    return this->operator=(temp op##= rhs); \
+}
+
     template<internal::NTTPString name, class T>
     struct InstanceProperty<name, T, true, true> {
-        explicit InstanceProperty(void* inst) noexcept : instance(inst) {}
+        explicit constexpr InstanceProperty(void* inst) noexcept : instance(inst) {}
         operator T() const {
             auto res = il2cpp_utils::GetPropertyValue<T, false>(reinterpret_cast<Il2CppObject*>(const_cast<void*>(instance)), name.data.data());
             if (!res) throw PropertyException(std::string("Failed to get instance property: ") + name.data.data());
             return *res;
         }
-        InstanceProperty& operator=(T&& t) {
+        InstanceProperty& operator=(T const& t) {
             auto val = reinterpret_cast<Il2CppObject*>(instance);
             auto res = il2cpp_utils::SetPropertyValue<false>(val, name.data.data(), std::forward<decltype(t)>(t));
             if (!res) throw PropertyException(std::string("Failed to set instance property: ") + name.data.data());
             return *this;
         }
+
+        inline auto operator ->() const {
+            return this->operator T();
+        } 
+
+        inline auto operator *() const {
+            return this->operator T();
+        }
+
+        inline auto v() const {
+            return this->operator T();
+        }
+
+        template<typename... Targs>
+        decltype(auto) operator ()(Targs&&... args) { return this->operator T()(std::forward<Targs...>(args...)); }
+
+        auto& operator ++() { return this->operator=(++this->operator T()); }
```

Comment:
Why aren't these `decltype(auto)`? Same with `BINARY_OPERATOR_OP_EQ_PROP`?

Reply:
the forward? because it wouldn't compile if I did that

Reply:
No, line 149 and below (the actual ops)

Reply:
![image](https://user-images.githubusercontent.com/51795403/197268720-9b4e6961-ca06-498e-a433-f16792632834.png)

Reply:
is this what you want?

Reply:
yeah